src/Sylius/Bundle/AssociationBundle/EventListener/LoadMetadataSubscriber.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\AssociationBundle\EventListener;
+
+use Doctrine\Common\EventSubscriber;
+use Doctrine\ORM\Event\LoadClassMetadataEventArgs;
+
+/**
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+class LoadMetadataSubscriber implements EventSubscriber
+{
+    /**
+     * @var array
+     */
+    protected $subjects;
+
+    /**
+     * @param array $subjects
+     */
+    public function __construct(array $subjects)
+    {
+        $this->subjects = $subjects;
+    }
+
+    /**
+     * @return array
+     */
+    public function getSubscribedEvents()
+    {
+        return [
+            'loadClassMetadata',
+        ];
+    }
+
+    /**
+     * @param LoadClassMetadataEventArgs $eventArgs
+     */
+    public function loadClassMetadata(LoadClassMetadataEventArgs $eventArgs)
+    {
+        $metadata = $eventArgs->getClassMetadata();
+        $metadataFactory = $eventArgs->getEntityManager()->getMetadataFactory();
+
+        foreach ($this->subjects as $subject => $class) {
+            if ($class['association']['classes']['model'] !== $metadata->getName()) {
+                continue;
+            }
+
+            $associationEntity = $class['subject'];
```

Comment:
I believe we can divide this foreach into three parts: mapping subject, mapping association and mapping association type.